File: content/examples/other/leetcode/problems/maximum-depth-of-binary-tree/example-2/solution.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def max_depth_bfs(root):
    if root is None:
        return 0
    q = deque([root])
    depth = 0
    while q:
        level_size = len(q)
        for _ in range(level_size):
            node = q.popleft()
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        depth += 1
    return depth


if __name__ == "__main__":
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    print(max_depth_bfs(root))

File: content/examples/other/leetcode/problems/maximum-depth-of-binary-tree/example-2/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function maxDepthBFS(root) {
  if (!root) return 0;
  const queue = [root];
  let depth = 0;
  for (let qi = 0; qi < queue.length; ) {
    const levelSize = queue.length - qi;
    for (let i = 0; i < levelSize; i += 1) {
      const node = queue[qi++];
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    depth += 1;
  }
  return depth;
}

if (require.main === module) {
  const root = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
  console.log(maxDepthBFS(root));
}

module.exports = { TreeNode, maxDepthBFS };
